(do not merge) - Sunset hdk docs and update to proc-macro (#1645)
holochain/holochain-rust: doc/holochain_101/src/first_steps.md

The `Serialize` and `Deserialize` derived traits allow the structs to be converted to and from JSON, which is how entries are managed internally in Holochain. The DefaultJson derived trait comes from the holochain HDK itself and allows for seamless converting between data stored in the DHT and rust structs.

> These structs on their own are not yet valid Holochain entries. To create an entry we must write a functio n that return a `ValidatingEntryType` and tag it using the `#[entry_def]` attribute. The `entry!` macro makes it easy to define a `ValidatingEntryType`.

Reviewer: ```suggestion
The List and ListItem structs on their own are not yet valid Holochain entries. To create an entry we must write a function that returns a `ValidatingEntryType` and tag it using the `#[entry_def]` attribute. The `entry!` macro makes it easy to define a `ValidatingEntryType`.
```

                Ok(())
            }
        )
    }
}
```

> Take note of the use of the struct types (`List` and `ListItem`) in the parameters to the validation function. This will automatically add validation to ensure these entries always match the structures. The `validation_package` field is a function that defines what data should be passed to the validation function through the `validation_data` argument. In this case we use a predefined function to only include the entry itself, but it is also possible to pass chain headers, chain entries or the full local chain. The validation field is a function that performs custom validation for the entry. In both our cases we are just returning `Ok(())`.

Reviewer: ```suggestion
Take note of the use of the struct types (`List` and `ListItem`) in the parameters to the validation function. This will automatically add validation to ensure these entries always match the structures, i.e. if a rouge node attempts to publish a List or ListItem that doesn't contain the data in the Struct definition, this will get caught by system level validation automatically.

The `validation_package` field is a function that defines what data should be passed to the validation function through the `validation_data` argument. In this case we use a predefined function to only include the entry itself, but it is also possible to pass chain headers, chain entries or the full local chain. The validation field is a function that performs custom validation for the entry. In both our cases we are just returning `Ok(())`.
```

Take note also of the `links` field. As we will see later links are the main way to encode relational data in holochain. The `links` section of the entry macro defines what other types of entries are allowed to link to and from this type. This also includes a validation function for fine grain control over linking.

## Adding Functions
Finally we need a way to interact with the hApp. We will define the following functions: `create_list`, `add_item` and `get_list`. get_list will retrieve a list and all the items linked to each list.

> Rust functions inside the zome module can be annotated with `#[zome_fn("hc_public")]` which will expose them to be callable from the conductor. The `"hc_public"` defines the capability required to call this function. This guide will not go in depth on capabilities but just know that hc_public means these functions are callable externally with no added security.

Reviewer: ```suggestion
Rust functions inside the zome module can be annotated with `#[zome_fn("hc_public")]` which will expose them to be callable from the conductor. The `"hc_public"` defines the capability required to call this function. This guide will not go in depth on capabilities but just know that hc_public means these functions are callable externally with no added security.  You can read more about capabilities [here](./zome/capabilities.md).
```
